Constructor selection based on @inject annotation, reformulated binding registration and circular dependency verification

Dreckr opened the pull request:
## Constructor selection based on @inject

The constructor is selected following this rules:

1- Is there any constructor annotated with @inject?
2- Is there only one constructor?
3- Is there a no-args constructor?

The first rule to pass is used to define the constructor. In case none of this rules pass, an exception is raised.
## Reformulated binding registration

Instead of creating provider functions for everything and having a map for singletons, now there is an actual Binding abstract class that defines a interface so that the instance construction and scope management logic can be abstracted from the Injector.
The main benefits of this new binding registration on this pull request is that constructors only have to be selected once for each binding (_ConstructorBinding stores the selected constructor so it can always use it without having to search for it again) and provider bindings can be verified for circular dependencies (_ProviderBinding stores the MethodMirror so it can know its parameters).
##  Circular dependency verification

After all the bindings were registered, they are analyzed to see if there is any circular dependency. This verification DOES NOT work in cases where the injector itself is injected (this includes bindTo cases, as the injector is injected silently into the module, as the constructor or provider can try to inject itself using it and there is no way to see this on the injectors initialization.

Dreckr committed:
Added @inject support

Constructors are now selected based on @inject annotation. If there is no constructor annotated with @inject, a no-args constructor is used. If there is no no-args constructor, an ArgumentError is thrown.

Fields annotated with @inject are injected after instance construction even if they are not set by the constructor. Final and Static fields with @injected annotation raise exceptions.

Support is not considered stable and tests will be implemented.

Dreckr committed:
Removed field injection and changed constructor selection a little bit

Dreckr committed:
Tests for constructor selection

The constructor is selected following this rules:

1- Is there any constructor annotated with @inject?
2- Is there only one constructor?
3- Is there a no-args constructor?

The first rule to pass is used to define the constructor. In case none of this rules pass, an exception is raised.

Dreckr committed:
README update

Dreckr committed:
Binding registration rewrite

The way that bindings are now stored keeps metadata information of the binding, like what constructor or provider is going to be used to create its instances. This will make it possible to analyze circular dependencies more easily.
API did not change at all as tests were left unchanged.

Dreckr committed:
Revert little undesired change in tests

Dreckr committed:
Removed trash from failed circular dependency detection attempt

Dreckr commented:
Closed for now, as more changes are coming

Dreckr committed:
Separated binding classes and circular dependency check.

Tests are failing because circular dependecies and unimplemented metadata on ParameterMirrors

Dreckr committed:
Tests for cyclical dependencies added and workaround for UnimplementedError on ParameterMirror.metadata

Dreckr committed:
updated docs

justinfagnani commented on lib/dado.dart:
```diff
@@ -59,11 +59,21 @@ library dado;
+
+Symbol _typeName(type) {
+  if (type is Type)
+    return reflectClass(type).qualifiedName;
```
style: put if statement blocks in braces unless they fit on one line

Dreckr committed:
Added braces on if blocks to follow Dart Style Guide

justinfagnani commented:
Awesome Diego, I'll take a look this weekend.

BTW, I hate reviews on Github. I'll try it out, but would you mind too much if we uploaded this to Reitveld so we can get side-by-side reviews? I does require installing Google's depot tools, but it integrates nicely with local git repos.

Dreckr commented:
Ok, I've never used it before, but I'll give it a try.

justinfagnani commented:
Thanks! Let me know if you have any trouble. The instructions are here: https://code.google.com/p/dart/wiki/PreparingYourMachine

Dreckr commented:
Here it is: https://codereview.appspot.com/11927045/

Only upload.py is needed though...

Dreckr committed:
Changes proposed by Adam

- Circular dependencies are verified by the injector, bindings only provide what are their dependencies;
- getSingletonOf is gone;
- Exception is raised when _getBinding is called and there is no binding for such key;
- _addBindingFromType renamed to _createBindingForType;
- Binding classes have been refactored to reuse code;
- New indirect circular dependency test with 2 levels of dependency;
- Minor changes.

Dreckr committed:
Code style fixes, _typeName is not polymorphic, improved Key.toString() and changed Key.annotation from dynamic to Object

Dreckr committed:
Object type annotation on metadata

Dreckr committed:
Minor changes

-Object type on annotations
-Print depedency stack on circular dependecy error
-Copyright notice
-Small bug fixes
-Cosmetic changes

Dreckr committed:
Removed unwanted evil code

Dreckr committed:
Made the injector manage injectors

Dreckr committed:
Fixes for Dart 1.0 and Mirrors API changes

justinfagnani commented:
Manually merged in https://github.com/dart-lang/dado/commit/cdb06e60dc1d649a6cacea1f01b2005dce3bc56c

Thanks Diego!!

Dreckr commented:
Awesome! I'm sorry I made this take so long, I'll make the next pull requests easier to merge.

Could you create issues for every feature and change you would like to see in Dado? I will have some spare time and I want to contribute a little more.
